torchx/workspace/docker_workspace.py

```diff
@@ -132,9 +144,24 @@ def build_workspace_and_update_role(
                 labels={
                     self.LABEL_VERSION: torchx.__version__,
                 },
+                decode=True,
             )
-            if len(old_imgs) == 0 or role.image not in old_imgs:
-                role.image = image.id
+            build_msg: Dict[str, str] = {}
+            while True:
```

Comment:
Can't we use a simple for loop?
```python
for build_msg in build_events:
    if "stream" in build_msg:
        output_str = build_msg["stream"].strip("\r\n").strip("\n")
        if output_str:
            log.info(output_str)
    if "aux" in build_msg:
        image_id = build_msg["aux"]["ID"]
        break
```

Reply:
build events is a stream, until stop_iteration is given we need to wait for further input on the stream. A for_loop wouldn't wait for the stream to complete and would only print events up to this point

Reply:
Isn't it how `Iterable`s/generators work in for loop in Python? If you give a for loop an `Iterable`/generator (something that has `__next__()`) it will call `__next__()` until it faces `StopIteration` and exists the loop.

Reply:
Seem to be working fine, please check the latest version of https://github.com/pytorch/torchx/pull/874